services/osparc-gateway-server/src/osparc_gateway_server/backend/osparc.py

```diff
@@ -182,3 +205,132 @@ async def do_check_workers(self, workers: List[Worker]) -> List[bool]:
+            #     cluster_update["config"] = unfrozen_cluster_config
+            max_workers = len(await get_cluster_information(self.docker_client))
+        if max_workers is not None and count > max_workers:
+            # This shouldn't happen under normal operation, but could if the
+            # user does something malicious (or there's a bug).
+            self.log.info(
+                "Cluster %s heartbeat requested %d workers, exceeding limit of %s.",
+                cluster_name,
+                count,
+                max_workers,
+            )
+            count = max_workers
+
+        if count != cluster.count:
+            cluster_update["count"] = count
+
+        created_workers = []
+        submitted_workers = []
+        target_updates = []
+        newly_running = []
+        close_expected = []
+        for worker in cluster.workers.values():
+            if worker.status >= JobStatus.STOPPED:
+                continue
+            elif worker.name in closing_workers:
+                if worker.status < JobStatus.RUNNING:
+                    newly_running.append(worker)
+                close_expected.append(worker)
+            elif worker.name in active_workers:
+                if worker.status < JobStatus.RUNNING:
+                    newly_running.append(worker)
+            elif worker.name in closed_workers:
+                target = (
+                    JobStatus.STOPPED if worker.close_expected else JobStatus.FAILED
+                )
+                target_updates.append((worker, {"target": target}))
+            else:
+                if worker.status == JobStatus.SUBMITTED:
+                    submitted_workers.append(worker)
+                else:
+                    assert worker.status == JobStatus.CREATED
+                    created_workers.append(worker)
+
+        n_pending = len(created_workers) + len(submitted_workers)
+        n_to_stop = len(active_workers) + n_pending - count
+        if n_to_stop > 0:
+            for w in islice(chain(created_workers, submitted_workers), n_to_stop):
+                target_updates.append((w, {"target": JobStatus.STOPPED}))
+
+        if cluster_update:
+            self.db.update_cluster(cluster, **cluster_update)
+            self.queue.put(cluster)
+
+        self.db.update_workers(target_updates)
+        for w, u in target_updates:
+            self.queue.put(w)
+
+        if newly_running:
+            # At least one worker successfully started, reset failure count
+            cluster.worker_start_failure_count = 0
+            self.db.update_workers(
+                [(w, {"status": JobStatus.RUNNING}) for w in newly_running]
+            )
+            for w in newly_running:
+                self.log.info("Worker %s is running", w.name)
+
+        self.db.update_workers([(w, {"close_expected": True}) for w in close_expected])
+
+
+async def _background_task(backend: OsparcBackend, cluster: Cluster):
```

Comment:
Are you using this at all? The call is commented out above and you call ```get_cluster_information``` already in the heartbeat callback.

Reply:
yes it's commented now with a todo for later